compiler/src/dmd/semantic3.d

```diff
@@ -270,6 +270,26 @@ private extern(C++) final class Semantic3Visitor : Visitor
         {
             if (funcdecl.storage_class & STC.inference)
             {
+                // https://issues.dlang.org/show_bug.cgi?id=23964
+                // The attributes of the generated opAssign are deduced
+                // by merging the attributes of the opAssign,
+                // postblit and destructor of the fields of the struct.
+                // (see dmd.clone.buildOpAssign).
+                // This does not take into account if any of the fields are
+                // system variables, because system variables are discovered
+                // during semantic2. As such, before analyzing the body
+                // of the generated opAssign, we need check if `sd` contains
+                // any systems variables and, if so, update the signature to
+                // be @system.
+                if (auto ad = funcdecl.isThis())
+                {
+                    if (auto sd = ad.isStructDeclaration())
+                    {
+                        if (sd.type.hasSystemFields())
+                            (cast(TypeFunction)funcdecl.type).trust = TRUST.system;
```

Comment:
Isn't this a breaking change, definitely infering `opAssign` as `@system` even without `-preview=systemvariables`?

Reply:
Using a system variable in a @safe contexts results in a deprecation without `-preview=systemvariables`. If you compile with `-de` that's gonna be an error. If this were special cased for `preview=systemvariables` then the reported bug would still manifest. Also, we are talking about inferring attributes on a function that uses system attributes. Even though accessing system variables only issues a deprecation, it seems to me that functions where the access is performed are inferred as system. Look at this example:

```d
  int fun() { return 1; }
  
void gun()()
{
      int a = fun();                                                                                                                                                                                                                              
}
  
void main() @safe
{
      gun();
}
```

This gives:

```
test.d(10): Error: `@safe` function `D main` cannot call `@system` function `test.gun!().gun`
test.d(1):        which calls `test.fun`
test.d(3):        `test.gun!().gun` is declared here
```

So, even though `int a = fun()` should only issue a deprecation, it actually affects the inference of attributes on `gun`. This was a breaking change that was accepted. Then why should the behavior for opAssign be any different?

Reply:
Calling `fun()`, a `@system` function, is what makes `gun()` `@system`, not variable `a`.

Reply:
My bad, it seems that moving `int a = fun()` outside of `gun` and then using a yields a deprecation. I can do the same for opAssign.